### PR TITLE
Refactors

### PR DESCRIPTION
**Pull request by mihaibivol**

### CODE REVIEW

File: README.rst [hunk 1]
```diff
@@ -41,9 +41,7 @@
         :target: https://github.com/inspirehep/json-merger/blob/master/LICENSE
 
 
-Invenio module that is able to merge json record objects.
-
-*This is an experimental developer preview release.*
+Module for merging JSON Objects.
```
Comment (kaplun) [inline README.rst:44]: We should amend also in setup.py where we say that it's an Invenio module.

File: json_merger/__init__.py [hunk 2]
```diff
@@ -22,17 +22,285 @@
 # waive the privileges and immunities granted to it by virtue of its status
 # as an Intergovernmental Organization or submit itself to any jurisdiction.
 
-"""Invenio module that is able to merge json record objects."""
+"""Module for merging JSON objects.
+
+To use this module, first import the main class:
```
Comment (kaplun) [inline json_merger/__init__.py]: ... you need to first import the main class:

File: json_merger/__init__.py [hunk 3]
```diff
@@ -22,17 +22,285 @@
...
+>>> from json_merger import Merger
+
+Then, import the configuration options:
+
+>>> from json_merger.config import UnifierOps, DictMergerOps
+
+The Basic Use Case
+------------------
+
+Let's assume we have JSON records in which don't have any list fields --
```
Comment (kaplun) [inline json_merger/__init__.py]: in which -> that

File: json_merger/__init__.py [hunk 4]
```diff
@@ -22,17 +22,285 @@
...
+... except MergeError:
+...     raised_something = True
+>>> m.merged_root == {
+...     'name': 'Jonathan',
+...     'age': 32,
+...     'address': 'Home',
+... }
+True
+
+If this type of conflict occurs, the merger will also populate a conflicts
```
Comment (kaplun) [inline json_merger/__init__.py]: conflicts -> `conflicts`

so that it's clear that it's a literal.

File: json_merger/comparator.py [hunk 5]
```diff
@@ -29,8 +29,17 @@
 
 
 class BaseComparator(object):
+    """Abstract base class for Entity Comparison."""
 
     def __init__(self, l1, l2):
+        """
+        Do not override the constructor since it will be called by the
+        json_merger logic.
```
Comment (kaplun) [inline json_merger/comparator.py]: Uh!? What do you mean?

File: json_merger/comparator.py [hunk 6]
```diff
@@ -68,21 +82,24 @@ class PrimaryKeyComparator(BaseComparator):
```
Comment (kaplun) [inline json_merger/comparator.py:79]: "Renders"? Maybe "Consider"?

**Review by kaplun — CHANGES_REQUESTED** [review]
What happened to configuration you had prepared so far? It could very well sit into an examples or contrib directory...

File: json_merger/comparator.py (hunk 5 above)
Comment (mihaibivol) [reply json_merger/comparator.py]: To leave the signatures correct, otherwise it will blow. A pretty useless comment actually :)

Comment (kaplun) [reply json_merger/comparator.py]: well then one can override the constructor as long as the signature is the same :P

Comment (mihaibivol) [reply json_merger/comparator.py]: :bike: :bike: :bike: Older logic was using self.l1 and self.l2. But whatever this is Python, devs should know what they do when they inherit from things. The example that I have shows you how to override two things.

Comment (kaplun) [reply json_merger/comparator.py]: Let me do my stupid comment any way. :stuck_out_tongue:  I can't really understand the whole rest of the refactoring unless I can focus on it for 24 hours in a row :stuck_out_tongue: 

Beside your "useless" docstring actually confused me, and that's why the comment.

So just put there: "this signature is expected by json-merger and should be preserved." or something

Comment (kaplun) [reply json_merger/comparator.py]: I mean... docstrings at this point are the only thing I can review. For the rest I assumes tests and APIs are the final source of truth.

Gaaah :boom:

**Comment by mihaibivol** [comment]
Logicwise, i did three things
- Moved some classes in dedicated files
- Redid the matching logic in graph_builder.
- Cache all matching pairs in comparators so any get matches call will be
  done in constant lookup.

Other than that I did doc tests. I should soon push the last bullets and
refactor a bit the API ones. Should be ready today :)

File: README.rst (hunk 1 above)
Comment (mihaibivol) [reply README.rst:44]: Done :)

**Comment by kaplun** [comment]
Good if you apply my small points, and squash the commits LGTM

**Review by kaplun — APPROVED** [review]

**Pull request merged**